pandas/core/groupby/indexing.py

```diff
@@ -0,0 +1,211 @@
+    FrameOrSeries,
+    PositionalIndexer,
+)
+from pandas.util._decorators import (
+    cache_readonly,
+    doc,
+)
+
+from pandas.core.dtypes.common import (
+    is_integer,
+    is_list_like,
+)
+
+if TYPE_CHECKING:
+    from pandas.core.groupby import groupby
+
+from pandas.core.indexes.api import CategoricalIndex
+
+
+class GroupByIndexingMixin:
+    """
+    Mixin for adding ._rows to GroupBy.
+    """
+
+    @property
+    def _rows(self) -> RowsGroupByIndexer:
+        if TYPE_CHECKING:
+            groupby_object = cast(groupby.GroupBy, self)
+        else:
+            groupby_object = self
+
+        return RowsGroupByIndexer(groupby_object)
+
+
+@doc(GroupByIndexingMixin._rows)
+class RowsGroupByIndexer:
+    def __init__(self, groupby_object: groupby.GroupBy):
+        self.groupby_object = groupby_object
+
+    def __getitem__(self, arg: PositionalIndexer | tuple) -> FrameOrSeries:
+        """
+        Positional index for selection by integer location per group.
+
+        Used to implement GroupBy._rows which is used to implement GroupBy.nth
+        in the case when the keyword dropna is None or absent.
+
+        Parameters
+        ----------
+        arg : PositionalIndexer | tuple
+            Allowed values are:
+            - int
+            - int valued iterable such as list or range
+            - slice with step either None or positive
+            - tuple of integers and slices
+
+        Returns
+        -------
+        Series
+            The filtered subset of the original groupby Series.
+        DataFrame
+            The filtered subset of the original groupby DataFrame.
+
+        See Also
+        --------
+        DataFrame.iloc : Purely integer-location based indexing for selection by
+            position.
+        GroupBy.head : Return first n rows of each group.
+        GroupBy.tail : Return last n rows of each group.
+        GroupBy.nth : Take the nth row from each group if n is an int, or a
+            subset of rows, if n is a list of ints.
+        """
+        with self.groupby_object._group_selection_context():
+            if isinstance(arg, tuple):
+                if all(is_integer(i) for i in arg):
+                    mask = self._handle_list(arg)
+                else:
+                    mask = self._handle_tuple(arg)
+
+            elif isinstance(arg, slice):
+                mask = self._handle_slice(arg)
+            elif is_integer(arg):
+                mask = self._handle_int(cast(int, arg))
+            elif is_list_like(arg):
+                mask = self._handle_list(cast(Iterable[int], arg))
+            else:
+                raise TypeError(
+                    f"Invalid index {type(arg)}. "
+                    "Must be integer, list-like, slice or a tuple of "
+                    "integers and slices"
+                )
+
+            ids, _, _ = self.groupby_object.grouper.group_info
+
+            # Drop NA values in grouping
+            mask &= ids != -1
```

Comment:
This is removing NA values after the determination of the mask - isn't this in conflict with nth?

> If dropna, will take the nth non-null row, dropna is either ‘all’ or ‘any’; this is equivalent to calling dropna(how=dropna) before the groupby.

Reply:
No, and this is confusing! There are:

1. The dropna df method mentioned here.
2. The dropna groupby parameter in e.g.  `df.groupby("A", dropna=False)`
3. The dropna nth parameter e.g. `gp.nth(3, dropby="any")`

2 refers to whether nans are dropped from the group column. That is the one being implemented here in code copied from nth (and maybe going back there. See Jeff's recent comment...)
3 refers to dropping rows based on nans in other columns and is the one in the nth docstring that you are referring to. That is separate code in nth that I have not touched (apart from tidying up the typing)

Reply:
Ah, right, thanks!